### PR TITLE
 feat(plugins): plugins can now define a bootstrap class

### PR DESCRIPTION
**Pull request by hypeJunction**

### CODE REVIEW

**Comment by hypeJunction** [comment]
This would solve a bunch of problems with testability and dev usability of plugins.

File: engine/classes/Elgg/Database/Plugins.php [hunk 1]
```diff
@@ -461,6 +461,9 @@ public function load() {
 
 		$this->hooks->getEvents()->registerHandler('plugins_boot:before', 'system', [$this, 'boot']);
 		$this->hooks->getEvents()->registerHandler('init', 'system', [$this, 'init']);
+		$this->hooks->getEvents()->registerHandler('ready', 'system', [$this, 'ready']);
+		$this->hooks->getEvents()->registerHandler('upgrade', 'system', [$this, 'upgrade']);
+		$this->hooks->getEvents()->registerHandler('shutdown', 'system', [$this, 'upgrade']);
```
Comment (hypeJunction) [inline engine/classes/Elgg/Database/Plugins.php]: `shutdown`

**Comment by hypeJunction** [comment]
@mrclay Any opinion?

**Comment by mrclay** [comment]
I dig it.

**Review by jdalsem — APPROVED** [review]

**Review by jeabakker — APPROVED** [review]

**Pull request merged**